### PR TITLE
disable anonymnous cipher suites to match Windows and OSX behavior

### LINKED ISSUE #23681: Inconsistent certificate handling between Linux and Windows
The following code yields different results on Linux and Windows (STARTTLS in SMTP)

On Windows the cert is handled correctly and the cert subj is printed to stdout.
On Linux (Fedora 26, Debian 9) cert is null in the CertificateValidationCallback (null if printed to stdout)

```c#
using System;
using Newtonsoft.Json;
using System.Net.Sockets;
using System.Net.Security;
using System.Security;
using System.IO;
using System.Security.Cryptography.X509Certificates;
using System.Security.Authentication;
using System.Text;
namespace code3
{
    class Program
    {
        static void Main(string[] args)
        {
            var host = "mail.galapp.net";
            using (var c = new TcpClient(host, 587))
            using (var s = c.GetStream())
            using (var sw = new StreamWriter(s, Encoding.ASCII))
            using (var sr = new StreamReader(s, Encoding.ASCII))
            {
                var l = read();
                sw.Write("EHLO localhost\r\n");
                sw.Flush();
                var l1 = read();
                sw.Write("STARTTLS\r\n");
                sw.Flush();
                var l2 = read();
                
                using (var ss = new SslStream(s, false, (snd, cert, chain, err) => {
                    Console.WriteLine(cert == null ? "null": cert.Subject);
                    return true;
                })) {
                    ss.AuthenticateAsClient(host, new X509CertificateCollection(), SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, false);                    
                }

                // handle SMTP response
                string read()
                {
                    byte[] buffer = new byte[1024];
                    int position = 0;
                    bool lastLine = false;

                    do
                    {
                        int readLength = s.Read(buffer, position, buffer.Length - position);
                        if (readLength > 0)
                        {
                            int available = position + readLength - 1;
                            if (available > 4 && (buffer[available] == '\n' || buffer[available] == '\r'))
                                for (int index = available - 3; ; index--)
                                {
                                    if (index < 0 || buffer[index] == '\n' || buffer[index] == '\r')
                                    {
                                        lastLine = buffer[index + 4] == ' ';
                                        break;
                                    }
                                }

                            position += readLength;

                            if (position == buffer.Length)
                            {
                                byte[] newBuffer = new byte[buffer.Length * 2];
                                Array.Copy(buffer, 0, newBuffer, 0, buffer.Length);
                                buffer = newBuffer;
                            }
                        }
                        else
                        {
                            break;
                        }
                    } while (!lastLine);

                    return Encoding.ASCII.GetString(buffer, 0, position - 1);
                }                    
            }
        }
    }
}
```

[EDIT] Add C# syntax highlighting by @karelz

### PR DESCRIPTION
**Pull request by wfurt**
The reason why this did not work and callback would have NULL certificate is that server did not sent any. TLS has obscure possibility to encrypt but not authenticate peer. With that, server would do anonymous DH key exchange instead of sending own certificate.  
This happens only when client (SSLStream in this case) offers such option and if server chooses it from the list. Windows do not have it in the list and that is why this happens to be platform difference. 

According to OpenSSL documentation:

> aNULL
> The cipher suites offering no authentication. This is currently the anonymous DH algorithms and anonymous ECDH algorithms. These cipher suites are vulnerable to a "man in the middle" attack and so their use is normally discouraged. These are excluded from the DEFAULT ciphers, but included in the ALL ciphers. Be careful when building cipherlists out of lower-level primitives such as kDHE or AES as these do overlap with the aNULL ciphers. When in doubt, include !aNULL in your cipher list.
> 

This change will take anonymous suite out of normal list used by default. When somebody forces less secure behavior, it would possibly stay in the list - that depends on OpenSSL compilation flags. 

Observer behavior only happens with specific combination of client/server. I did not find good way how to write unit test - since we do not have api to get or set cipher list. If anybody has suggestions, please let me know. 

Fixes #23681

### CODE REVIEW

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.cpp [hunk 1]
```diff
@@ -510,15 +510,17 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
 // delimiter ":" is used to allow more than one strings
 // below string is corresponding to "AllowNoEncryption"
 #define SSL_TXT_Separator ":"
+#define SSL_TXT_Exsclusion "!"
```
Comment (rmkerr) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.cpp]: nit: exclusion

File: src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.cpp [hunk 2]
```diff
@@ -510,15 +510,17 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
...
 #define SSL_TXT_AllIncludingNull SSL_TXT_ALL SSL_TXT_Separator SSL_TXT_eNULL
+#define SSL_TXT_NotAnon SSL_TXT_Separator SSL_TXT_Exsclusion SSL_TXT_aNULL
```
Comment (rmkerr) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.cpp]: Same as above.

**Comment by rmkerr** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop UWP CoreCLR x64 Debug Build

**Comment by davidsh** [comment]
>I also set up HTTP server allowing only anonymous handshake and I verified that SocketHttpClient fails to connect with authentication exception.

You should document exactly how you did this so that someone else can repeat the manual test.  Perhaps a docker file, using openssl script etc.

**Comment by wfurt** [comment]
I added ssl.cipher-list = "aNULL:" to my lighttpd conf. The simplest way is probably something like:

> $ openssl s_server -www -cert server.pem -cipher aNULL

That should negotiate _only_ anonymous cipher suites - on platforms where it is supported. 

Any more thoughts on this @bartonjs ?

**Comment by bartonjs** [comment]
> Any more thoughts on this @bartonjs ?

Nope :smile:.

I agree that `openssl s_server -www -cert server.pem -cipher aNULL` is likely the easiest way to set up a test.

**Pull request merged**